modules/auxiliary/admin/dcerpc/icpr_cert.rb

```diff
@@ -61,7 +62,24 @@ def run
   end
 
   def action_request_cert
-    request_certificate
+    with_ipc_tree do |opts|
+      request_certificate(opts)
+    end
   end
 
+  # @yieldparam options [Hash] If a SMB session is present, a hash with the IPC tree present. Empty hash otherwise.
+  # @return [void]
+  def with_ipc_tree
+    opts = {}
+    if session
+      print_status("Using existing session #{session.sid}")
+      client = session.client
+      self.simple = ::Rex::Proto::SMB::SimpleClient.new(client.dispatcher.tcp_socket, client: client)
+      opts[:tree] = simple.client.tree_connect("\\\\#{client.dispatcher.tcp_socket.peerhost}\\IPC$")
+    end
+
+    yield opts
+  ensure
+    opts[:tree].disconnect! if opts[:tree]
```

Comment:
Not a blocker for me; but it looks like the current request_ceritificate/connect_ipc method doesn't perform IPC cleanup if it opens up a new connection

https://github.com/rapid7/metasploit-framework/blob/401cdca09b61dfb6fcf920284e1aabf7477519c8/lib/msf/core/exploit/remote/ms_samr.rb#L165

In ruby_smb's [examples](https://github.com/rapid7/ruby_smb/blob/f29b6811854fb6bf20a5d9069c7205c89601983d/examples/list_directory.rb#L32) there are other places that aren't cleaned up correctly

It also looks like as part of tree_connect there's an array of `@tree_connects`:

https://github.com/rapid7/ruby_smb/blob/f29b6811854fb6bf20a5d9069c7205c89601983d/lib/ruby_smb/client.rb#L614

But the caching doesn't get cleaned up when you later call `tree.disconnect!`:

https://github.com/rapid7/ruby_smb/blob/f29b6811854fb6bf20a5d9069c7205c89601983d/lib/ruby_smb/smb1/tree.rb#L43

It also looks like there's two ways to disconnect trees, via the tree instance, or via the client instance. Neither cleanup the @tree_connects instance 

https://github.com/rapid7/ruby_smb/blob/f29b6811854fb6bf20a5d9069c7205c89601983d/lib/ruby_smb/client/utils.rb#L69-L71

Nothing actionable from this, just documenting my findings